lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2980,26 +3030,97 @@ class PrimalGenCloner final
+        errorOccurred = true;
+        return;
+      }
+    }
+    // Check and diagnose non-differentiable results.
+    if (!originalFnTy->getResults()[indices.source]
+            .getSILStorageType()
+            .isDifferentiable(context.getModule())) {
+      context.emitNondifferentiabilityError(
+          original, invoker, diag::autodiff_nondifferentiable_result);
       errorOccurred = true;
       return;
     }
-    auto vjp = vjpAndVJPIndices->first;
+    // If VJP has not yet been found, emit an `autodiff_function` instruction
+    // on the remapped original function operand and `autodiff_function_extract`
+    // the VJP. The actual JVP/VJP functions will be populated in the
+    // `autodiff_function` during the transform main loop.
+    if (!vjp) {
+      // FIXME: Handle indirect differentiation invokers. This may require some
+      // redesign: currently, each original function + attribute pair is mapped
+      // only to one invoker.
+      /*
+      DifferentiationInvoker indirect(ai, this->original, attr);
+      auto insertion =
+          context.getInvokers().try_emplace({this->original, attr}, indirect);
+      auto &invoker = insertion.first->getSecond();
+      invoker = indirect;
+      */
+
+      // If the original `apply` instruction has a substitution map, then the
+      // applied function is specialized.
+      // In the VJP, specialization is also necessary for parity. The original
+      // function operand is specialized with a remapped version of same
+      // substitution map using an argument-less `partial_apply`.
+      if (!ai->getSubstitutionMap().empty()) {
+        auto substMap = getOpSubstitutionMap(ai->getSubstitutionMap());
+        auto vjpPartialApply = getBuilder().createPartialApply(
+            ai->getLoc(), original, substMap, {},
+            ParameterConvention::Direct_Guaranteed);
+        original = vjpPartialApply;
+      }
+
+      auto *autoDiffFuncInst = getBuilder().createAutoDiffFunction(
+          loc, indices.parameters, /*differentiationOrder*/ 1, original);
+
+      // Record the `autodiff_function` instruction.
+      context.getAutoDiffFunctionInsts().push_back(autoDiffFuncInst);
+      context.getResultIndices()[autoDiffFuncInst] =
+          activeResultIndices.front();
+
+      vjp = getBuilder().createAutoDiffFunctionExtract(
+          loc, AutoDiffFunctionExtractInst::Extractee::VJP,
+          /*differentiationOrder*/ 1, autoDiffFuncInst);
+    }
 
     // Record desired/actual VJP indices.
     // Temporarily set original pullback type to `None`.
-    NestedApplyInfo info{indices, vjpAndVJPIndices->second,
-                         /*originalPullbackType*/ None};
-    auto insertion = getDifferentiationTask()->getNestedApplyInfo().try_emplace(
-        ai, std::move(info));
-    assert(insertion.second && "Nested apply info already exists");
+    NestedApplyInfo info{indices, indices, /*originalPullbackType*/ None};
```

Comment:
`desiredIndices` and `actualIndices` should be removed from `NestedApplyInfo`, now that `emitAssociatedFunctionReference` is no longer called within `visitApply`.

`actualIndices` are resolved in `ADContext::processAutoDiffFunctionInst`.

Reply:
Done in https://github.com/apple/swift/pull/24845/commits/10522e5882b8c1e0714a9523827abc2a653a0187. Also removed original `SILFunction *` from `SILDifferentiableAttribute` `DifferentiationInvoker` case.